contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,251 @@
+
+        pub let nodeID: String
+
+        pub let phase: UInt8
+
+        pub let content: String
+
+        init(nodeID: String, phase: UInt8, content: String) {
+            self.nodeID = nodeID
+            self.phase = phase
+            self.content = content
+        }
+    }
+
+    // The Participant resource is generated for each consensus node when they register.
+    // Each resource instance is good for all future potential epochs, but will
+    // only be valid if the node operator has been confirmed as a consensus node for the next epoch.
+    pub resource Participant {
+
+        pub let nodeID: String
+
+        // Submits a whiteboard message to the contract
+        pub fun sendMessage(phase: UInt8, _ content: String) {
+            pre {
+                FlowDKG.participantIsRegistered(self.nodeID): "Cannot send whiteboard message if not registered for the current epoch"
+                phase > FlowDKG.currentPhase && phase < UInt8(4): "Phase submission is invalid"
+            }
+
+            // create the message struct
+            let message = Message(nodeID: self.nodeID, phase: phase, content: content)
+
+            // add the message to the message record for the phase
+            FlowDKG.phaseMessages[phase]!.append(message)
+
+            emit BroadcastMessage(phase: phase, nodeID: self.nodeID, content: content)
+
+        }
+
+        /// Sends the final key vector submission. 
+        /// Can only be called during the final phase and by consensus nodes that are registered.
+        pub fun sendFinalSubmission(_ submission: [String]) {
+            pre {
+                FlowDKG.participantIsRegistered(self.nodeID): "Cannot send final submission if not registered for the current epoch"
+                !FlowDKG.nodeHasSubmittedFinal(self.nodeID)!: "Cannot submit a final submission twice"
+                FlowDKG.currentPhase == DKGPhase.finalSubmission: "Can only send final submission in the final DKG phase"
+                // submission should be a certain length?
+            }
+
+            var finalSubmissionIndex = 0
+
+            // iterate through all the existing unique submissions
+            // If this participant's submission matches one of the existing ones,
+            // add to the counter for that submission
+            // Otherwise, track the new submission and set its counter to 1
+            for existingSubmission in FlowDKG.uniqueFinalSubmissions {
+                var index = 0
+                // If the submission length is different than the one being compared to
+                // move on to the next one
+                if submission.length != existingSubmission.length {
+                    return
+                }
+
+                // Check each key in the submiission to make sure that it matches
+                // the existing one
+                for key in submission {
+                    if key.length != 128 {
+                        // correct key length?
```

Comment:
I'll need to confirm, and this will depend on the encoding scheme we use when submitting the transaction. Maybe add a constant for this with a TODO for now

Reply:
Any update on the key length? @jordanschalm @tarakby

Reply:
what key are we considering here @joshuahannan ?

Reply:
[This is the list of public keys in the final submission](https://www.notion.so/dapperlabs/Smart-contract-based-DKG-29ad04d8996c4aae8019348ce49085a0#3da54491bbf54578bb0e8b384aa42c01)

Reply:
A single public key is 96 bytes. So following the document the total is `96 * (D+1)` bytes.
However, the keys would be all `nil` if the dkg node detected that the protocol failed.

Reply:
👍 Added this to the notion: https://www.notion.so/dapperlabs/Smart-contract-based-DKG-29ad04d8996c4aae8019348ce49085a0#e6e411af2cea452a94a55159de791a6f

> However, the keys would be all nil if the dkg node detected that the protocol failed.

We consider the DKG protocol failing to be a un-recoverable failure currently, so I'm OK to simply not handle this case in the contract for now